flow/layers/display_list_raster_cache_item.cc

```diff
@@ -0,0 +1,185 @@
+    // in doing to extra work to rasterize.
+    return false;
+  }
+
+  if (display_list == nullptr ||
+      !RasterCacheUtil::CanRasterizeRect(display_list->bounds())) {
+    // No point in deciding whether the display list is worth rasterizing if it
+    // cannot be rasterized at all.
+    return false;
+  }
+
+  if (is_complex) {
+    // The caller seems to have extra information about the display list and
+    // thinks the display list is always worth rasterizing.
+    return true;
+  }
+
+  unsigned int complexity_score = complexity_calculator->Compute(display_list);
+  return complexity_calculator->ShouldBeCached(complexity_score);
+}
+
+DisplayListRasterCacheItem::DisplayListRasterCacheItem(
+    DisplayList* display_list,
+    const SkPoint& offset,
+    bool is_complex,
+    bool will_change)
+    : RasterCacheItem(RasterCacheKeyID(display_list->unique_id(),
+                                       RasterCacheKeyType::kDisplayList),
+                      CacheState::kCurrent),
+      display_list_(display_list),
+      offset_(offset),
+      is_complex_(is_complex),
+      will_change_(will_change) {}
+
+std::unique_ptr<DisplayListRasterCacheItem> DisplayListRasterCacheItem::Make(
+    DisplayList* display_list,
+    const SkPoint& offset,
+    bool is_complex,
+    bool will_change) {
+  return std::make_unique<DisplayListRasterCacheItem>(display_list, offset,
+                                                      is_complex, will_change);
+}
+
+void DisplayListRasterCacheItem::PrerollSetup(PrerollContext* context,
+                                              const SkMatrix& matrix) {
+  cache_state_ = CacheState::kNone;
+  DisplayListComplexityCalculator* complexity_calculator =
+      context->gr_context ? DisplayListComplexityCalculator::GetForBackend(
+                                context->gr_context->backend())
+                          : DisplayListComplexityCalculator::GetForSoftware();
+
+  if (!IsDisplayListWorthRasterizing(display_list_, will_change_, is_complex_,
+                                     complexity_calculator)) {
+    // We only deal with display lists that are worthy of rasterization.
+    return;
+  }
+
+  transformation_matrix_ = matrix;
+  transformation_matrix_.preTranslate(offset_.x(), offset_.y());
+
+  if (!transformation_matrix_.invert(nullptr)) {
+    // The matrix was singular. No point in going further.
+    return;
+  }
+
+  if (context->raster_cached_entries && context->raster_cache) {
+    context->raster_cached_entries->push_back(this);
+    cache_state_ = CacheState::kCurrent;
+  }
+  return;
+}
+
+void DisplayListRasterCacheItem::PrerollFinalize(PrerollContext* context,
+                                                 const SkMatrix& matrix) {
+  if (cache_state_ == CacheState::kNone || !context->raster_cache ||
+      !context->raster_cached_entries) {
+    return;
+  }
+  if (!context->raster_cache->GenerateNewCacheInThisFrame()) {
```

Comment:
Ignore the access_threshold for now. It's a way to turn off the cache which is only used for testing so we don't care if we test it early or late, just leave it in GNCITF() and test it during the TryToPrepare method and it will work just fine for both purposes.

Reply:
Done

Reply:
I've got it